godtossing/Pos/Ssc/GodTossing/Toss/Base.hs

```diff
@@ -144,78 +199,54 @@ computeSharesDistrPure
 computeSharesDistrPure richmen threshold
     | null richmen = pure mempty
     | otherwise = do
-        let total :: Word64
-            total = sum $ map unsafeGetCoin $ toList richmen
-        when (total == 0) $
+        when (totalCoins == 0) $
             throwError $ TossInternallError "Richmen total stake equals zero"
-        let epsilon = sharesDistrInaccuracy::Rational
         -- We accept error in computation = 0.05,
         -- so stakeholders must have at least 55% of stake (for reveal secret) in the worst case
         let mpcThreshold = toRational (getCoinPortion threshold) / toRational coinPortionDenominator
-        let fromX = 1
+        let fromX = ceiling $ 1.0 / minimum portions
         let toX = truncate $ toRational (3::Int) / mpcThreshold
 
         let keys = map fst $ HM.toList richmen
-        let portions = map ((`divRat` total) . unsafeGetCoin . snd) $ HM.toList richmen
-        unless (all (>= mpcThreshold) portions) $
+        unless (all ((>= mpcThreshold) . toRational) portions) $
             throwError $ TossInternallError "Richmen stakes less than threshsold"
 
-        let init = normalize $ multPortions portions toX
-        let initS = sum init
-        let initDelta = calcSumError (map (`divRat` initS) init) portions
-        (_, _, res) <-
-                execStateT (compute fromX toX epsilon portions) (initDelta, initS, init)
-        pure $ HM.fromList $ zip keys res
+        case compute fromX toX 0 of
+            Nothing -> throwError $ TossInternallError "There is no appropriate distribution"
+            Just res -> pure $ HM.fromList $ zip keys res
   where
+    sharesDistrInaccuracyR = sharesDistrInaccuracy :: Double
+
+    coins :: [CoinUnsafe]
+    coins = map (fromIntegral . unsafeGetCoin) (toList richmen)
```

Comment:
Hm, previously you didn't want to rely on the fact that `HM.toList`, `toList`, `HM.elems` and `HM.keys` all return the same order.
Now you rely on this fact here. So probably it makes sense to use `HM.keys` above.